File: 448_find_all_numbers_disappeared_in_an_array/Solution2/Solution.py
class Solution:
    def findDisappearedNumbers(self, nums: list) -> list:
        if not nums: return []

        result = []
        for num in nums: nums[abs(num) - 1] = -abs(nums[abs(num) - 1])
        for i, num in enumerate(nums):
            if num > 0: result.append(i + 1)

        return result


if __name__ == '__main__':
    nums = [4, 3, 2, 7, 8, 2, 3, 1]
    print(Solution().findDisappearedNumbers(nums))

File: 448_find_all_numbers_disappeared_in_an_array/Solution2/Solution.java
import java.util.*;

class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        List<Integer> result = new ArrayList<>();
        if (nums == null || nums.length == 0) return result;

        for (int num : nums) nums[abs(num) - 1] = -abs(nums[abs(num) - 1]);
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > 0) result.add(i + 1);
        }

        return result;
    }

    public int abs(int num) {
        return num >= 0 ? num : -num;
    }

    public static void main(String[] args) {
        int[] nums = {4, 3, 2, 7, 8, 2, 3, 1};
        System.out.println(new Solution().findDisappearedNumbers(nums));
    }
}
